<h1><b>Statistique en Bioinformatique : </b> TME7 </h1>
<br>
L’objectif de ce TME est: 
<br>
<ul>
<li> Construire un modèle d'une famille proteique avec HMM profil. Ce modèle facilite la reconnaissance des membres de la famille, en particulier les protéines les plus distantes (point de vue évolutif). </li> 
</ul>
<br>
<div class="alert alert-warning" role="alert" style="margin: 10px">
<p><b>Soumission</b></p>
<ul>
<li>Renomer le fichier TME7.ipynb pour NomEtudiant1_NomEtudiant2.ipynb </li>
<li>Soumettre via moodle </li>
</div>
</div>


Nom etudiant 1 :
<br> Anastasia AZMOUDEH
Nom etudiant 2 :
<br> Késia DETHELOT-DELAG 

---
<h3> Preparation => Installer HMMer 2.3.2 </h3>

---

Nous allons travailler avec deux versions du logiciel HMMer: 2.3.2 et 3. Installez les deux versions dans deux répertoires différents. Tout d'abord, nous allons considérer la version 2.3.2<br><br>

<h4>1- Installer la version 2.3.2</h4><br>
Télécharger HMMer <a href="http://eddylab.org/software/hmmer/2.3/hmmer-2.3.tar.gz">version 2.3.2</a> <br>
Lisez les instructions et installez-le dans votre espace personnel. <br><br>

<font face="monospace">tar -xzf hmmer-2.3.tar.gz</font><br>
<font face="monospace">cd hmmer-2.3</font><br>
<font face="monospace">./configure --prefix=/home/yourUser/yourDirectory/hmmer-2.3</font><br>
<font face="monospace">make</font><br>

<font color="red"> Remplacez yourUser/yourDirectory par vos coordonnées personnels.</font>

<h4>2 - Dans le terminal modifier le fichier .bashrc</h4> 

<font face="monospace">cd \$HOME </font><br>
<font face="monospace">gedit .bashrc </font><br>
Ajouter la ligne suivante à votre fichier <font face="monospace">.bashrc</font><br>
<font face="monospace">export PATH=/home/yourUser/yourDirectory/hmmer-2.3/src:${PATH}</font><br>

<font color="red"> Remplacez <font face="monospace">yourUser/yourDirectory</font> par votre coordonnée personnelle.</font><br>
Sauvergardez le fichier et taper dans le terminal <br>

<font face="monospace">cd $HOME</font><br>
<font face="monospace">source .bashrc </font><br>




<b>Exercice 1 </b>: <u>Zinc finger Pfam domain</u> <br>
<b>1.1</b>: Construire le modèle HMMer pour le domaine Pfam Zinc finger, numéro d'accès PF00096. Pour cela, téléchargez l'alignement seed au format Stockholm depuis le site web de <a href="http://pfam.xfam.org/family/PF00096#tabview=tab3">Pfam</a>. Construire le modèle HMMer avec les paramètres par défaut, enregistrer avec le nom PF00096.hmm

<b>1.2</b>: Utilisez le programme HMMeditor pour visualiser le modele PF00096.hmm et répondre aux questions.<br>
<b>a)</b>: Combien d'états Match observez-vous?

Reponse: 
<font color="blue">
 On observe 26 états de Match.
</font>

<b>b)</b>: Pourquoi les états M3, M8, M12, M18, M21 et M26 ont une bordure en gras?

Reponse:
<font color="blue">
A ces états, on observe une forte conservation pour un acide aminé donné.
</font>




<b>c)</b>: Quelle est la probabilité de transition M5-> I5? 

Reponse:
<font color="blue">
La probabilité est de 0.06.
</font>

<b>d)</b>: Quel est l'acide aminé le plus fréquent dans l'état I5?



Reponse: 
<font color="blue">
Les acides aminés les plus fréquents à I5 sont G (glycine) et S (sérine)  avec une probabilité de 0.09 chacun. 
</font>

<b>1.3</b>: Télécharger la séquence X6NX52 à <a href="http://www.uniprot.org/uniprot/X6NX52">Uniprot</a>. Utilisez le programme <font face="monospace">hmmsearch</font> pour localiser les domaines  Zinc finger dans la séquence X6NX52. Combien de domaines obtenez-vous? Comparez vos résultats avec les résultats de  <a href="http://pfam.xfam.org/protein/X6NX52_RETFI">Pfam</a>. Avez-vous trouvé les mêmes domaines dans les mêmes positions? Expliquez votre réponse.

Reponse: 
<font color="blue"><br>
Avec le programme hmmsearch, on trouve 4 domaines aux positions 74-97, 101-125, 143-166 et 221-244. <br>
Avec Pfam, on ne trouve que le domaine à la position 221-241, il est raccouci de 3 acides aminés par rapport aux résultats de hmmersearch. <br>
Ces différences s'expliquent par le fait que la recherche avec hmmersearch est plus permissive, donc on détecte plus de domaines. 
    
<font>

<b>1.4</b>: Enregistrer la séquence du premier domaine, positions 74-97, dans un format fasta avec le nom de fichier dom1.fasta, supprimer les gaps de la séquence avant de l'enregistrer. Chargez le fichier dom1.fasta dans le programme HMMeditor et cliquez sur "Align Path". Qu'observez-vous? 

Reponse: 
<font color="blue"> <br> On observe le chemin des états.
    

</font>




<b>1.5</b>: Modifiez le fichier dom1.fasta et ajouter le mot RWV après le mot TFR. Rechargez le fichier dans le programme HMMeditor et clique sur "Align Path". Qu'observez vous?

Reponse:<font color="blue"><br>
En ajoutant le mot RWW, on constate qu'on passe par état I (Insert) supplémentaire par rapport au HMM de la question 1.4
Ce résultat est logique car on a ajouté un mot de 3 par rapport à la séquence question 1.4 .
</font>


---
<h3> Preparation => Installer HMMer 3 </h3>

---

Télécharger HMMer <a href="http://eddylab.org/software/hmmer3/3.1b2/hmmer-3.1b2.tar.gz">version 3</a> <br>
Lisez les instructions et installez-le dans votre espace personnel. <br>

<font face="monospace">tar -xzf hmmer-3.1b2.tar.gz</font><br>
<font face="monospace">cd hmmer-3.1b2</font><br>
<font face="monospace">./configure --prefix=/home/yourUser/yourDirectory/hmmer-3.1b2</font><br>
<font face="monospace">make</font><br>

<font color="red"> Remplacer yourUser/yourDirectory avec votre donnée personnelle.</font>

<h4>-Dans le terminal modifier le fichier .bashrc</h4>

<font face="monospace">cd \$HOME </font><br>
<font face="monospace">gedit .bashrc </font><br>
Mettre en commentaire le export de hmmer-2.3 (utiliser # au début de la ligne) et ajouter la ligne suivante à votre fichier .bashrc<br>
<font face="monospace">export PATH=$PATH:/home/yourUser/yourDirectory/hmmer-3.1b2/src;</font>
<font color="red"> Remplacer yourUser/yourDirectory avec votre donnée personnelle.</font>
Sauvegarder le fichier .bashrc et dans le terminal taper<br>
<font face="monospace">source .bashrc </font><br>

<b>Exercice 2 </b>: <u>Ubiquitin domain Pfam domain</u> <br>
<b>2.1</b>: Construire le modèle HMMer pour le domaine Ubiquitin Pfam, accession number PF00240. Pour cela, téléchargez l'alignement seed au format Stockholm depuis le site web de  <a href="http://pfam.xfam.org/family/PF00240#tabview=tab3">Pfam</a>. Construire le modèle hmmer avec les paramètres par défaut, enregistrer sous le nom PF00240.hmm. Combien d'états Match observez-vous?


Reponse: 
<font color="blue">
<br> L'architecture comporte 72 élements. On sait que l'architecture définit le nombre de match, on observe donc 72 match.  

</font>


<b>2.2</b>:
Téléchargez toutes les séquences avec le domaine ubiquitin du site Pfam. Clique sur l'alignement, sur Format an alignment, sélectionnez Full, Format FASTA,  Gaps : No gaps unaligned. Enregistrez le fichier sous le nom PF00240_full.fasta. Combien de séquences observez-vous dans le fichier PF00240_full.fasta?

In [1]:
#Read the fasta file
def read_fasta (input_f):
    """
    Read a fasta file
    input1 input_f: fasta file name
    output1 arr_seqs: a two dimension array containing SeqID and amino acide sequence
    """
    arr_seqs = []
    
    names = [] # contient les noms des séquences du fichier FASTA
    sequences = [] # contient les séquences présentes dans le fichier FASTA
    
    f = open(input_f)
    une_seq = []
    
    for lines in f.readlines():
        
        if lines[0]== ">":
            sequences.append(une_seq) 
            une_seq = []
            if lines[-1]=="\n": 
                names.append(lines[1:-1])
            else: 
                names.append(lines[1:])
                
        else:
            if lines[-1]=="\n": 
                une_seq.append(lines[:-1])
            else: 
                une_seq.append(lines)
            

    sequences.append(une_seq)
    sequences = sequences[1:] # pour supprimer la première liste qui est vide
    
    nvlles_seq = []
    for liste in sequences: 
        longueur = len(liste)
        liste = [''.join(liste[:longueur])] # pour qu'une même séquence écrite sur 2 lignes du fichier FASTA soit dans un seul item de la liste 
#         print(liste)
        nvlles_seq.append(liste)
    
    

    arr_seqs.append(nvlles_seq) 
    arr_seqs.append(names)
    
    return arr_seqs 



input_PF00240_full = 'PF00240_full.fasta'
sequences_PF00240_full, names_PF00240_full = read_fasta (input_PF00240_full)

print(len(sequences_PF00240_full))



21402


Reponse: <font color="blue"> <br>
Dans le fichier PF00240_full.fasta, on observe 21 402 séquences. 
</font>

<b>2.3</b>:
Utilisez hmmsearch (default parameters) pour détecter les domaines ubiquitin dans le fichier PF00240_full.fasta. Utilisez l'option --domtblout pour enregistrer un fichier tabulaire, donnez le nom PF00240.domtb, répondre aux questions suivantes:

In [2]:
#Read the domtblout file
def read_domtblout(input_f):
    """
    load the results of domtblout file
    input1 input_f: domtblout file name
    output hits: a matrix containg hit information
    """
   
    hits = []
    f=open(input_f)
    for lines in f.readlines():
        
        if lines[0]=='#':
            continue
            
        else:
            hits.append(list(lines.split()))
            
    return hits 

input_domtblout = 'PF00240bis.domtb'
hits_PF00240 = read_domtblout(input_domtblout)
print (hits_PF00240[0])
#['R0KPY3_NOSB1/3-74', '-', '72', 'PF00240', '-', '72', '5.2e-36', '122.4', '0.5', '1', '1', '5.7e-36', '5.7e-36', '122.3', '0.5', '1', '72', '1', '72', '1', '72', '0.99', '-\n']

['R0KPY3_NOSB1/3-74', '-', '72', 'PF00240', '-', '72', '5.2e-36', '122.4', '0.5', '1', '1', '5.7e-36', '5.7e-36', '122.3', '0.5', '1', '72', '1', '72', '1', '72', '0.99', '-']


<b>a)</b>: Combien de domaines observez-vous dans le fichier PF00240.domtb? Comparez le nombre de domaines détectés avec le nombre attendu (question 2.2)

In [3]:
#Some code if necessary
len(hits_PF00240)

21435

Reponse: 
<font color="blue"> <br>
Dans le fichier PF00240.domtb, on détecte 21 435 domaines. D'après la question 2.2, on s'attendait à 21 402 domaines. On a donc détecté 33 domaines en plus que les domaines attendus. On conjecture  que l'option domtblout permet de détecter plus précisément les doamines. 
</font>

<b>b)</b>: Quelle est le c-evaleus et les scores du meilleur et du pire hit?

Reponse:
<font color="blue"><br>
La c-Evalue du pire hit est 5.2e-05.
<br> La c-Evalue du pire meilleur est 5.7e-36.

</font>

In [4]:
import numpy as np
Hits = np.asarray(hits_PF00240) # on transforme en np.array pour ensuite pouvoir sélectionner une colonne

Evalues = Hits[:,6] # on sélectionne la colonne des e-values

Evalues=Evalues.astype(float) # on s'assure que tous les chiffres soient des float et pas des strings pour les comparer ultérieurement
# print(Evalues)

mini=min(Evalues)
maxi=max(Evalues)

# print("MIN = ",min(Evalues))
# print("MAX = ",max(Evalues))

# print("Np.where")
# print(np.where(Evalues==maxi))
# print(np.where(Evalues==mini))

ind_maxi=np.where(Evalues==maxi)
# print(ind_maxi[0][0])
ind_worst = ind_maxi[0][0]
print("c-Evalue du pire hit est :",hits_PF00240[ind_worst][11])

ind_mini=np.where(Evalues==mini)
ind_best=ind_mini[0][0]
print("c-Evalue du pire meilleur est :",hits_PF00240[ind_best][11])

c-Evalue du pire hit est : 5.2e-05
c-Evalue du pire meilleur est : 5.7e-36


<b>c)</b>: Exécutez à nouveau hmmsearch et définissez le domain e-value à 1e-5. Combien de domaines observez-vous? Comparez le nombre de domaines détectés avec l'attendu (question 2.2) et obtenu dans la question a. Justifiez votre réponse.

In [5]:
#Some code if necessary
input_domE = 'PF00240eval.domtb'
hitsEval_PF00240 = read_domtblout(input_domE)
len(hitsEval_PF00240)

21340

Reponse:
<font color="blue"> <br>
Cette fois-ci, on trouve 21 340 domaines avec un threshold e-value de 1e-5, c'est 95 domaines de moins que sans threshold (question a). En utilisant un threshold, on ne retient pas toutes les données. La quantité de domaines trouvés correspond toujours à un peu plus de la moitié des domaines attendus. 

<font>


<b>d)</b>:
Téléchargez le modèle hmm pour le domaine ubiquitin depuis le site de Pfam. Clique sur "Curation & model" et sur "download the raw HMM for this family". Exécutez à nouveau hmmsearch avec l'option --cut_ga. Comparez le nombre de domaines détectés avec les questions précédentes. Qu'est-ce que cut_ga fait?

In [6]:
input_cutga = 'PF00240cutga.domtb'
hitscutga_PF00240 = read_domtblout(input_cutga)
print(len(hitscutga_PF00240)) # 21 449

# input_cutgab = 'PF00240cutgabis.domtb' # document téléchargé en ligne  
# hitscutgab_PF00240 = read_domtblout(input_cutgab)
# print(len(hitscutgab_PF00240)) # 21 332

#--cut_ga : use profile's GA gathering cutoffs to set all thresholding


21449


Reponse: 
<font color="blue">
<br>
 On trouve, cette fois-ci, 21 449 domaines au lieu de 21435. L'option cut_ga permet de trouver des domaines que l'on n'avait pas détectés la première fois. En effet, cette option utilise le score de bits GA pour définir les seuils d'inclusion et de reporting par séquences et domaines. Or, les seuils GA sont plus fiables pour définir l'appartenance familiale. 
<font>

<b>Exercice 3 </b>: <u>Cooccurrence de domaines</u> <br>
Le domaine Ribosomal_S27 (PF01599) est très suivant trouvé avec le domaine de l'ubiquitine?  Allez sur le site Pfam, recherchez le domaine PF01599 et cliquez sur "Domain Organisation" recherchez le domaine ubiquitin dans la liste des architectures de domaine. Répondre aux question suivantes:<br>
<b>3.1</b>:
Si ces domaines coexistent, dans combien de séquences cette architecture de domaines est-elle observée? 

Reponse:
<font color="blue"><br>
Les deux domaines Ribosomal_S27 (PF01599) et le domaine de l'ubiquitine coexistent dans 3 509 protéines. 
<font>

<b>3.2</b>: Vérifions si la séquence W4GKD7 décrite comme protéine ribosomale S27a de l'Ubiquitine-40S a l'architecture du domaine: ubiquitine, Ribosomal_S27. <br>
- Téléchargez la séquence de W4GKD7 de <a href="http://www.uniprot.org/uniprot/W4GKD7">Uniprot</a>  et téléchargez le modèle Ribosomal_S27 hmmer à partir du site Web Pfam.
- Concattez les modèles ubiquitin.hmm et Ribosomal_S27.hmm, enregistrez le fichier sous Ribosomal_S27_ubiquitin.hmm.
- Exécutez hmmsearch avec l'option --cut_ga pour rechercher des domaines dans W4GKD7, utilisez le modèle de fichier Ribosomal_S27_ubiquitin.hmm. Enregistrez la sortie au format tabulaire, utilisez l'option --domtblout pour enregistrer un fichier tabulaire, nommé W4GKD7.domtb
- Analyser le fichier W4GKD7.domtb, avez-vous détecté les domaines Ribosomal_S27 et ubiquitin? Quels sont les scores et les e-values?

Reponse:
<font color="blue"> <br> 
Les scores sont de:
<br> -->  360.7 pour le domaine W4GKD7
<br> -->  91.2 pour le domaine Ribosomal_S27 <br>
<br> Les e-values sont de: 
<br> -->  3.8e-113  pour le domaine W4GKD7
<br> -->  1.5e-30 pour le domaine Ribosomal_S27 <br>
<font>


<b>3.3</b>: Visualisez l'architecture du domaine pour la protéine W4GKD7 en utilisant le programme  <a href="http://prosite.expasy.org/cgi-bin/prosite/mydomains/">myDomains</a>; Remplissez la longueur de la protéine et les données de domaine. Vous pouvez trouver toutes les informations dans le fichier W4GKD7.domtb, choisir la forme et la couleur que vous préférez, inclure l'image dans votre reponse.

Reponse: <font color="blue"> 
<br> Le domaine PF00240 se situe à la position 3-74. <br> Le domaine Ribosomal_S27 se situe à la poisiton 102-147.
<font>
<img src="PSImage.png" alt="Drawing" style="width: 300px;"/>

<b>Exercice 4 </b>: <u>Domaines du même clan</u><br>
Les familles de Pfam ayant des fonctions similaires sont suivant orgnisés dans un clan, c'est le cas des familles ubiquitin et ubiquitin_2 placé dans le clan CL0072.<br>
<b>4.1</b>: Combien de membre exists dans le clans CL0072?

Reponse:<font color="blue"> <br>
Dans le clan CL0072, on compte 79 membres.
<font>

<b>4.2</b>: Téléchargez le modèle hmm pour le domaine ubiquitin_2 depuis le site de Pfam. Clique sur "Curation & model" et sur "download the raw HMM for this family". Concatenez les modeles ubiquitin.hmm et ubiquitin_2.hmm and executez hmmsearch avec --cut_ga sur les sequences du fichier Uniprot_ubiquitin.fasta, utilizez l'option --domtblout pour enregistrer un fichier tabulaire. Repondrez aux questions:


<b>a)</b>: Combien de sequences ont plus qu'un hit?

In [7]:
# une séquence a plusieurs hits si son nom apparait plusieurs fois dans la 
# colonne target name du fichier Uniprot_ubiquitin.domtb

# fichier = 'res.domtb'
fichier = 'Uniprot_ubiquitin.domtb'
lecture = read_domtblout(fichier)
# print(lecture[0])


liste_noms = [] # colonne target name du fichier Uniprot_ubiquitin.domtb 

for i in range (len(lecture)): 
    liste_noms.append(lecture[i][0]) 

#print(noms)
# print(len(noms))


plusieurs = [] # liste contenant les séquences avec plusieurs hits
for i in  (liste_noms): 
    nb = liste_noms.count(i)
    
    if nb > 1:
        if i not in plusieurs: 
            plusieurs.append(i)
        else: 
            continue

        
            
print("Réponse :",len(plusieurs), "séquences contiennent plusieurs hits")
# print(plusieurs)


Réponse : 37 séquences contiennent plusieurs hits


Reponse: <font color="blue"> 
37 séquences ont plusieurs hits.
<font>

<b>b)</b>: Faite une fonction pour filtrer les meilleurs hits de protéines membre de ubiquitin à partir d'un fichier domtb. 
Si une séquence a plus qu'un hit dans de positions chevauchantes vous devez choisir le hit avec le meilleur c-evalue. 
Le début et fin de chaque hit se trouve dans les colonnes 20 et 21 du fichier de sortie de hmmsearch.

In [14]:
input_exo4 = 'Uniprot_ubiquitin.domtb'
hits_exo4 = read_domtblout(input_exo4)

In [22]:
def best_hits(hits, eValue_thresh, overlap_thresh):
    """
    Filtering the best hits
    input1 hits: hit table (output of domtb file)
    input2 eValue_thresh: E-value Threshold (default 1e-100)
    input3 overlap_thresh: overlaping Threshold (default 10)
    output1 meilleur_hits: filtered hit table 
    """
    meilleur_hits = {}
    hits_a_supp ={}
    hits_dico={}
    
    for e in hits_exo4:
        if e[0]+"-"+e[2]+"-"+e[3] not in hits_dico:
            hits_dico[e[0]+"-"+e[2]+"-"+e[3] ]=[(e[11],(e[19],e[20]))]
        else:
            hits_dico[e[0]+"-"+e[2]+"-"+e[3]].append((e[11],(e[19],e[20])))

    dico_res={}
    res_l={}
    for key,val in hits_dico.items():
        
        for e in val:
            ev,pos=e
            
            if key not in res_l:
                res_l[key]=[pos]
            else:
                res_l[key].append(pos)
                
    dicoch={}
    for key,val in res_l.items():
        
        for i in range(1,len(val)):
            if int(val[i-1][1])> int(val[i][0]) or int(val[i-1][0]) > int(val[i][0]):
                
                if key not in dicoch:
                    dicoch[key]=[(i-1,i)]
                else:
                    dicoch[key].append((i-1,i))
                    
#    print("DICO CHEVAUCHEMENT ",dicoch)
    for key,val in dicoch.items():
#         print("POUR LA SEQ ",key)
        for e in val:
            i,j=e
#             print("les chevauchements sont sur les sous séquences numéros ")
#             print("i = ",i)
#             print("j = ",j)
#             print("pour i ",hits_dico[key][i])
#             print("pour j",hits_dico[key][j])
            evli,posi=hits_dico[key][i]
            evlj,posj=hits_dico[key][j]
            if float(evli) < float(evlj):
                #print("La c-evalue de i ",i," (",evli,") est meilleure que celle de j ",j,"(", evlj, ")")
                if key not in meilleur_hits:
                    meilleur_hits[key]=[i]
                    hits_a_supp[key]=[j]
                else:
                    meilleur_hits[key].append(i)
                    hits_a_supp[key].append(j)
                
            else:
                #print("la c-evalue de j",j,"(",evlj,") est meilleure que celle de i",i,"(",evli,")" )
                if key not in meilleur_hits:
                    meilleur_hits[key]=[j]
                    hits_a_supp[key]=[i]
                else:
                    meilleur_hits[key].append(j)
                    hits_a_supp[key].append(i)
    
    print("Remarque: Il faut supprimer les hits suivants\n", hits_a_supp, '\n')
                                     
    return meilleur_hits

                                     
input_fasta = 'Uniprot_ubiquitin.fasta'
sequences,seq_names = read_fasta (input_fasta)

input_domtblout = 'Uniprot_ubiquitin.domtb'
hits_ubiquitin = read_domtblout(input_domtblout)

eValue_thresh = 8e-100
overlap_thresh = 10

meilleur_hits = best_hits(hits_ubiquitin, eValue_thresh, overlap_thresh)
print("MEILLEUR_HITS\n",meilleur_hits)

Remarque: Il faut supprimer les hits suivants
 {'tr|A0A0V0UP34|A0A0V0UP34_9BILA-1182-W4GKD7': [13, 13], 'tr|A0A0V0TBY2|A0A0V0TBY2_9BILA-919-W4GKD7': [8], 'tr|U6KZM6|U6KZM6_EIMTE-1189-W4GKD7': [1, 1, 2, 4, 4, 5, 7, 7, 9, 10, 10, 12, 12, 13, 15, 15, 17, 17, 18, 20, 20, 21, 22, 23, 25, 25, 26, 28, 28, 30, 30, 31, 33, 33, 34, 36, 36, 38, 38], 'tr|A0A0V0UP34|A0A0V0UP34_9BILA-1182-Ubiquitin_2': [1]} 

MEILLEUR_HITS
 {'tr|A0A0V0UP34|A0A0V0UP34_9BILA-1182-W4GKD7': [12, 14], 'tr|A0A0V0TBY2|A0A0V0TBY2_9BILA-919-W4GKD7': [9], 'tr|U6KZM6|U6KZM6_EIMTE-1189-W4GKD7': [0, 2, 3, 3, 5, 6, 6, 8, 8, 9, 11, 11, 13, 14, 14, 16, 16, 18, 19, 19, 21, 22, 23, 24, 24, 26, 27, 27, 29, 29, 31, 32, 32, 34, 35, 35, 37, 37, 39], 'tr|A0A0V0UP34|A0A0V0UP34_9BILA-1182-Ubiquitin_2': [2]}
